pydarn/plotting/rtp.py

```diff
@@ -0,0 +1,438 @@
+                                                dmap_record[parameter][j]
+                                else:
+                                    z[i][dmap_record['slist'][j]] = \
+                                            dmap_record[parameter][j]
+                                    # calculate min and max value
+                                    if z[i][dmap_record['slist'][j]] < z_min or\
+                                       z_min is None:
+                                        z_min = z[i][dmap_record['slist'][j]]
+                                    if z[i][dmap_record['slist'][j]] > z_max:
+                                        z_max = z[i][dmap_record['slist'][j]]
+                    # a KeyError may be thrown because slist is not created
+                    # due to bad quality data.
+                    except KeyError:
+                        continue
+
+        # Check if there is any data to plot
+        if np.all(np.isnan(z)):
+            raise rtp_exceptions.RTPNoDataFoundError(parameter, beam_num,
+                                                     start_time, end_time)
+        time_axis, elev_axis = np.meshgrid(x, y)
+        z_data = np.ma.masked_where(np.isnan(z.T), z.T)
+
+        if color_norm is None:
+            norm = colors.Normalize(z_min, z_max)
+        else:
+            norm = color_norm(z_min, z_max)
+
+        cmap = cm.get_cmap(settings['color_map'])
+        # set ground scatter to grey
+        cmap.set_under('grey', 1.0)
+
+        # plot!
+        im = ax.pcolormesh(time_axis, elev_axis, z_data, lw=0.01,
+                           cmap=cmap, norm=norm)
+        # setup some standard axis information
+        ax.set_xlim([start_time, end_time])
+        ax.xaxis.set_major_formatter(dates.DateFormatter(settings['date_fmt']))
+        ax.yaxis.set_ticks(np.arange(0, y_max+1, 15))
+        ax.xaxis.set_minor_locator(dates.HourLocator())
+        ax.yaxis.set_minor_locator(ticker.MultipleLocator(5))
+        # so the plots gets to the ends
+        ax.margins(0)
+
+        # create color bar if True
+        cb = None
+        if settings['color_bar']:
+            cb = ax.figure.colorbar(im, ax=ax)
+            cb.set_label(settings['color_bar_label'])
+
+        return im, cb, cmap
+
+    @classmethod
+    # TODO: could parallelize this method
+    def __filter_data_check(cls, dmap_record: dict, settings: dict, j: int) -> bool:
+        """
+        checks for data that does not meet the criteria of the filtered
+        settings
+
+        Parameters
+        ----------
+        dmap_record : dict
+            dictionary of the dmap record fields
+        settings : dict
+            dictionary of the settings list
+        j : int
+            index on the slist or range gate to check the array values
+
+        Returns
+        -------
+        pass_flg : bool
+            boolean indicating if the dmap_record passes all filter checks
+        """
```

Comment:
Recommend adding some comments here above each for-loop that says what it is doing.